### PR TITLE
refactor(compiler): rework defer block analysis

### PR DESCRIPTION
**Pull request by crisbeto**
Currently we have the `deferrableDeclToImportDecl`, `deferBlocks`, `deferrableTypes` and `deferBlockDepsEmitMode` fields on the `R3ComponentMetadata` which is incorrect, because the interface is used both for JIT and AOT mode even though the information for those fields is AOT-specific. It will be problematic for partial compilation since the runtime will have a reference to the dependency loading function, but will not be able to provide any of the other information.

These changes make the following refactors:
1. It changes the defer-related information in `R3ComponentMetadata` to include only references to dependency functions which can be provided both in JIT and AOT.
2. Moves the AOT-specific defer analysis into the `ComponentResolutionData`.
3. Moves the construction the defer dependency function into the compilation phase of the `ComponentDecoratorHandler`.
4. Drops support for defer blocks from the `TemplateDefinitionBuilder`. This allows us to clean up some TDB-specific code and shouldn't have an effect on users since the TDB isn't used anymore.

### CODE REVIEW

File: packages/compiler/src/render3/view/template.ts [hunk 1]
```diff
@@ -1327,192 +1323,7 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
...
-
-      // The most common case should be a trigger within the view so we can omit a depth of
-      // zero. For triggers in parent views and in the placeholder we need to pass it in.
-      if (depth !== 0) {
-        params.push(o.literal(depth));
-      }
-
-      return params;
-    });
+    throw new Error('Deferred blocks no longer supported in TemplateDefinitionBuilder');
```
Comment (crisbeto) [inline packages/compiler/src/render3/view/template.ts:1326]: Context here is that for the `TemplateDefinitionBuilder` to work, we would have to resolve the triggers during analysis. We already have the logic, but given that this isn't necessary in the template pipeline and that the pipeline is the default, I figured it was a good time to clean up the logic.

File: packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts [hunk 2]
```diff
@@ -1483,23 +1471,77 @@ export class ComponentDecoratorHandler implements
...
-      triggerElements: Map<TmplAstDeferredTrigger, TmplAstElement|null>): void {
-    Object.keys(triggers).forEach(key => {
-      const trigger = triggers[key as keyof TmplAstDeferredBlockTriggers]!;
-      triggerElements.set(trigger, componentBoundTarget.getDeferredTriggerTarget(block, trigger));
-    });
+  private compileDeferBlocks(resolution: Readonly<Partial<ComponentResolutionData>>):
+      R3DeferMetadata {
+    if (resolution.deferBlockDepsEmitMode === DeferBlockDepsEmitMode.PerBlock) {
+      if (!resolution.deferBlockDependencies) {
+        throw new Error('deferBlockDependencies must be present when compiling in PerBlock mode');
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:1485]: nit: we can mention here (and in a similar `throw`s within this function) that this is an internal error (i.e. unrelated to developers code):
```suggestion
        throw new Error('Internal error: deferBlockDependencies must be present when compiling in PerBlock mode');
```

File: packages/compiler/src/render3/view/template.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/compiler/src/render3/view/template.ts:1326]: I'd propose adding more info into the error message to make sure that it's not interpreted as "defer is no longer supported", here is a proposal:
```suggestion
    throw new Error('Deferred blocks are no longer supported by the obsolete TemplateDefinitionBuilder, ' +
      'Please enable the new template compilation pipeline for your application.');
```

File: packages/compiler/src/template/pipeline/src/ingest.ts [hunk 3]
```diff
@@ -443,9 +442,13 @@ function ingestDeferView(
...
 
 function ingestDeferBlock(unit: ViewCompilationUnit, deferBlock: t.DeferredBlock): void {
-  const blockMeta = unit.job.deferBlocksMeta.get(deferBlock);
-  if (blockMeta === undefined) {
-    throw new Error(`AssertionError: unable to find metadata for deferred block`);
+  let ownResolverFn: o.ArrowFunctionExpr|null = null;
+
+  if (unit.job.deferMeta.mode === DeferBlockDepsEmitMode.PerBlock) {
+    if (!unit.job.deferMeta.blocks.has(deferBlock)) {
+      throw new Error(`AssertionError: unable to find a depdendencies fuction for deferred block`);
```
Comment (AndrewKushnir) [inline packages/compiler/src/template/pipeline/src/ingest.ts:449]:
```suggestion
      throw new Error(`AssertionError: unable to find a dependency function for this deferred block`);
```

File: packages/compiler/src/template/pipeline/src/phases/resolve_defer_deps_fns.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    for (const op of unit.create) {
+      if (op.kind === ir.OpKind.Defer) {
+        if (op.resolverFn !== null) {
+          continue;
+        }
+
+        if (op.ownResolverFn !== null) {
+          if (op.handle.slot === null) {
+            throw new Error(
+                'AssertionError: slot must be assigned bfore extracting defer deps functions');
```
Comment (AndrewKushnir) [inline packages/compiler/src/template/pipeline/src/phases/resolve_defer_deps_fns.ts:26]:
```suggestion
                'AssertionError: slot must be assigned before extracting defer deps functions');
```

**Review by AndrewKushnir — APPROVED** [review]
@crisbeto great refactoring, thank you 👍 Just left a few minor comments.

**Comment by crisbeto** [comment]
Feedback has been addressed. Will mark this for merge once the RC goes out.

**Comment by atscott** [comment]
This PR was merged into the repository by commit eee620aa0013cdb6e860f402d2064c50ada46d4e.

**Pull request closed without merging**